### ISSUE TITLE
EH debugging: reloading EH window does not restart debug session

### ISSUE THREAD

**Issue by weinand** [Microsoft/vscode#29334]
Insiders 1.14.0:

- open "Hello World" extension project in VS Code
- F5 to start extension 
- press "Restart" button

Observe: EH window reloads and debug session is restarted. This works fine.

- in Hello World extension reload window or open another project

Observe: EH window reloads but debug session terminates. This is the bug.

I've debugged this on the node-debug side and I see that the `restart` attribute of the `DebugProtocol.DisconnectArguments` is `true` for the 'Restart' case and 'false' for the Reload and Open case. It should be 'true' for all three cases.

**Comment by weinand** [comment]
Without a fix EH debugging becomes pretty useless because you cannot open the workspace required for the extension that you need to debug.

**Comment by isidorn** [comment]
@bpasero looked into this and I am not getting an event in debug for the reload of the workbench. More precise [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/debugService.ts#L165). This might have been introduced with some multi root refactoring. I am investigating further to find the root cause

**Comment by isidorn** [comment]
@weinand I am not sure that VS Code was properly sending the `restart` attribute to true in the past.
In the order for the `restart` argument to be passed the debug adapter should terminate with an event that has `restart` in the `body` specifeid [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/common/debugProtocol.d.ts#L127)


What I am seeing on the vscode side is that once a user reloads vscode
1. Debug adapter terminates -> I get a `terminate` event ( I am not sure who inititated this terminate)
2. Because I get a regualr `terminate` event I send a `disconnect` without the `restart` flag to the adapter

I believe there is some timing issue in the Insiders, since I can not repro the problem out of source, but can repro in insiders.

Bottom line: I think the root cause is not missing of the restart flag but something else (timing issue)

**Comment by weinand** [comment]
I've investigated this a bit more and found this:
- "Reload Window" was already broken in 1.13 and 1.12, but "Open" was working fine so we did not notice this problem.
- the debugger UI was never sending a `"restart": "true"` in the "Disconnect" request so it seems that we were always relying on a specific timing (and in Insider the timing has changed it a way that now even the "Open" triggers the issue).

I will experiment on the DA side to fix this. The problem is that the DA cannot send a terminate event with a `"restart": "true"` because it does not know that the EH terminated because of a "Reload" or "Open" (which both would need the restart flag to become true).

**Comment by weinand** [comment]
This problem started in VS Code 1.12 und it seems to be independent from the DA.

If I run VS Code 1.11.2 against the latest node-debug it works fine, whereas VS Code 1.12.2 against the latest node-debug shows the problem.

**Comment by weinand** [comment]
```
24191: __initialize: got process id 24189 from node
24191: _initialize: target node version: v7.4.0
24191: _injectDebuggerExtensions: code injection successful
24191: _startInitialize: stopped: true
24191: _startInitialize: no entry event after 10 retries; giving up
24191: NodeV8Protocol hook: ref cache size: 10
24191: _startInitialize2: fire initialized event
24191: _startInitialize2: remember to do a 'Continue' later
24191: setBreakPointsRequest: {"path":"/Users/weinand/MS/Projects/Scenarios/extensions/hello-ext/extension.js","name":"extension.js"} [{"line":20}]
24191: _setBreakpoint: {"type":"scriptRegExp","target":"^(.*[\\/\\\\])?\\/Users\\/weinand\\/MS\\/Projects\\/Scenarios\\/extensions\\/hello-ext\\/extension\\.js$","line":19,"column":0}
24191: _updateBreakpoints: result [{"verified":true,"line":20,"column":1}]
24191: setFunctionBreakPointsRequest: result []
24191: setExceptionBreakPointsRequest: ["uncaught"]
24191: configurationDoneRequest: do a 'Continue'
24191: _stopped: got exception event from node
24191: _stopped: new ref cache
```
here I pressed Cmd-R for reloading the window. What follows might be a hint towards the problem. 
An "attach" request is sent to the old debug adapter which is already shutting down instead of the new DA.
```
24191: attachRequest: args: {"port":11321}
24191: _attach: address: 127.0.0.1 port: 11321
24191: _terminated: socket end
24191: _terminated: node v8protocol close
24191: shutdown: kill debugee and sub-processes
24191: disconnectRequest: send response
```

**Comment by isidorn** [comment]
Thanks to git bisect we have a winner @jrieken commit https://github.com/Microsoft/vscode/commit/876c35b8cd9e95c3e8ca041f40257b5f1c657028

It seems like because the renderer got faster(?) the timing of events I am getting from @bpasero changed. I can investigate more into the exact timing of events, however we alrady released one stable version with this since this commit is 10weeks old so it might be ok to ship with this issue and devise a better strategy for extension debugging in july.
In any case I will look more into the ordering of events and will check if there is an easy fix.

![screen shot 2017-06-29 at 17 09 22](https://user-images.githubusercontent.com/1926584/27695145-5eff307e-5cee-11e7-8fb9-ae57f0ce8654.png)

**Comment by jrieken** [comment]
Yeah, that commit is very old and we have shipped with that. If this is actually timing dependent, then trigger reload window which will *not* use cached data, adding roughly 700-800ms to the startup (loading workbench.js).

**Comment by weinand** [comment]
Yes, we have this problem since VS Code 1.12, but initially it affected only "Reload Window" and not "Open Workspace". In more recent releases additional performance improvements seem to have worsen the timing issue and now it affects "Open Workspace" as well.

@bpasero for this milestone could we disable using cached data when debugging?

**Comment by bpasero** [comment]
@weinand you need to convince @jrieken

**Comment by weinand** [comment]
@jrieken @bpasero 
We (the debugger) are not using "Reload Window" or "Open Workspace". So in order to get the old timing behaviour, we need to turn off the data cache for a renderer under debugging.
Is this easily doable?

**Comment by jrieken** [comment]
You need to convince @egamma and co. Tho I am not how that fits into this

> "Reload Window" was already broken in 1.13 and 1.12, but "Open" was working fine so we did not notice the problem.

"Reload Window" is the case in which cached-data *isn't used* (node/electron limitation). So this is the slow case. Cached-data *is used* on new windows, meaning that it is the fast case. So, I wonder what it helps to make the fast case the a slow case, which isn't working since long(?). Also, I'd say the "Reload Window" is more important then "Open New Window" when debugging an extension. I might be missing what we are aiming for...

If you wanna turn off cached data, do it [here](https://github.com/Microsoft/vscode/blob/master/src/vs/platform/environment/node/environmentService.ts#L137) for renderers and [here](https://github.com/Microsoft/vscode/blob/master/src/main.js#L217) for the main process.

**Comment by isidorn** [comment]
I have verified that turning off cached data for the renderer fixes the issue.

However for now I am simply reacting on the extension attach event with a 300 ms timeout, which gives the right ordering of events. I randomly picked 300ms and this works on my machine, however not sure how will other machines behave.
The real fix would be not to be dependent on the order of events and this we can tackle in the future.

I verified on mac, it would be great if the verifier tries on win and linux.

Right order after user reloads:
extension host dies, node adapter terminates
event from workbench to extensionAttach

This ordering was the other way around. Not sure if we could solve this in a nicer way directly from where we emit events @bpasero 
Though in the end I believe we would also do some timeout strategy which I currently did on the debug side.

**Comment by egamma** [comment]
@isidorn can you please file a new issue so that we do not forget to come with a better fix for this than to rely on time outs.

**Comment by weinand** [comment]
@isidorn great, this confirms my hypothesis from above that the rawAttach was coming too early (and hitting the wrong session).
I'll verify the workaround tomorrow for linux and Windows.
Thanks for helping me with this.

**Comment by isidorn** [comment]
@egamma created https://github.com/Microsoft/vscode/issues/29884 so we can tackle this properly 
@weinand welcome

**Comment by bpasero** [comment]
Sorry, I can still reproduce. When I reload the EH window (macOS), the debug toolbar disappears indicating to me that debugging connection is not re-established.

Hard to reproduce, maybe harder than before, but hitting it sometimes.

**Comment by bpasero** [comment]
I am seeing very weird behaviour now: I reload and when a breakpoint is hit, my EH debugging window closes? Is that a new regression?

**Comment by weinand** [comment]
macOS: "opening a workspace" works fine again. I never saw a dropped debug session. "reload window" still drops the session occasionally. But workaround is simple: just repeat the "reload window".

**Comment by bpasero** [comment]
The first time I reload the EH, the debug toolbar disappears as shown below. But it comes back when I reload again:

![flicker_chrome58](https://user-images.githubusercontent.com/900690/27730027-29b06df8-5d88-11e7-8499-e08db53efe83.gif)

**Comment by weinand** [comment]
Since the current situation has much improved over the last stable (1.13) I suggest that we us the fix for 1.14.

**Comment by isidorn** [comment]
Closing this issue as @weinand mentioned the current behavior is better than the last release. As for the real fix we have #29884 to track it
